components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/AssignUserRolesFunctionImpl.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.CarbonException;
+import org.wso2.carbon.core.util.AnonymousSessionUtil;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.JsAuthenticatedUser;
+import org.wso2.carbon.identity.application.authentication.framework.exception.FrameworkException;
+import org.wso2.carbon.identity.conditional.auth.functions.user.internal.UserFunctionsServiceHolder;
+import org.wso2.carbon.user.core.UserRealm;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+
+import java.util.List;
+
+/**
+ * Function to update given roles for a given user.
+ * The purpose is to perform role assigning during dynamic authentication.
+ */
+public class AssignUserRolesFunctionImpl implements AssignUserRolesFunction {
+
+    private static final Log LOG = LogFactory.getLog(AssignUserRolesFunctionImpl.class);
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param user           Authenticated user.
+     * @param assigningRoles Roles to be assigned.
+     * @return <code>true</code> If the role assigning is successfully completed. <code>false</code> for any other case.
+     */
+    @Override
+    public boolean assignUserRoles(JsAuthenticatedUser user, List<String> assigningRoles) {
+
+        boolean isDebugEnabled = LOG.isDebugEnabled();
+
+        if (user != null && assigningRoles != null) {
+            try {
+                if (user.getWrapped() != null) {
+                    String tenantDomain = user.getWrapped().getTenantDomain();
+                    String userStoreDomain = user.getWrapped().getUserStoreDomain();
+                    String username = user.getWrapped().getUserName();
+                    UserRealm userRealm = getUserRealm(tenantDomain);
+                    if (userRealm != null) {
+                        UserStoreManager userStore = getUserStoreManager(tenantDomain, userRealm, userStoreDomain);
+                        userStore.updateRoleListOfUser(
+                                username,
+                                new String[0],
+                                assigningRoles.toArray(new String[0])
+                        );
+                        return true;
+                    } else if (isDebugEnabled) {
+                        LOG.debug("Unable to find userRealm for the user: "
+                                + username + " in userStoreDomain: " + userStoreDomain);
+                    }
+                } else if (isDebugEnabled) {
+                    LOG.debug("Unable to get wrapped content for the user");
+                }
+            } catch (UserStoreException e) {
+                LOG.error("Error while getting user from the store", e);
+            } catch (FrameworkException e) {
+                LOG.error("Error while retrieving userRealm and userStoreManager", e);
+            }
+        } else {
+            if (user == null) {
```

Comment:
will be much clear if we do the individual validation at the top of the method and log/return on error.